codec/gcm.go

```diff
@@ -29,23 +61,42 @@ func NewGCMSession(key [32]byte) (*GCMSession, error) {
 	if err != nil {
 		return &GCMSession{}, fmt.Errorf("creating gcm cipher: %v", err)
 	}
-	return &GCMSession{
-		gcm:       gcm,
-		wRand:     rand.New(rand.NewSource(int64(binary.BigEndian.Uint64(key[:8])))),
-		wBuf:      make([]byte, gcm.NonceSize()),
-		nonceSize: gcm.NonceSize(),
-	}, nil
+
+	gcmSession := &GCMSession{
+		gcm:        gcm,
+		readNonce:  &gcmNonce{},
+		writeNonce: &gcmNonce{},
+	}
+
+	if bytes.Compare(self[:], remote[:]) < 0 {
+		gcmSession.writeNonce.top = math.MaxUint32
+		gcmSession.writeNonce.bottom = math.MaxUint64
+		gcmSession.writeNonce.countDown = true
+	} else {
+		gcmSession.readNonce.top = math.MaxUint32
+		gcmSession.readNonce.bottom = math.MaxUint64
+		gcmSession.readNonce.countDown = true
```

Comment:
Both of these cases start with the exact same settings. I believe one of them is meant to be the complement of the other.

Reply:
It's correct, I think. One branch is setting the writeNonce value to all 1s while the other one is doing it to readNonce.

Reply:
Oh right, of course. Completely misread that.